File: backend/routers/smart_catalog.py
from typing import Optional

from fastapi import APIRouter, HTTPException, Query

from data.smart_catalog import (
    CATALOG_CATEGORIES,
    CATALOG_VEHICLES,
    get_hotspots,
    get_vehicle,
    search_hotspots,
)
from database import get_conn, row_to_product
from models import CatalogHotspot, CatalogHotspotProduct, CatalogVehicle, CatalogVehicleSummary, Product


router = APIRouter(prefix="/api/smart-catalog", tags=["smart-catalog"])


@router.get("/categories")
def list_categories():
    return CATALOG_CATEGORIES


@router.get("/vehicles", response_model=list[CatalogVehicleSummary])
def list_vehicles(search: Optional[str] = Query(None)):
    items = CATALOG_VEHICLES
    if search:
        q = search.strip().lower()
        items = [v for v in items if q in v["name"].lower() or q in v.get("subtitle", "").lower()]
    return [
        CatalogVehicleSummary(
            id=v["id"],
            name=v["name"],
            subtitle=v.get("subtitle", ""),
            year=v.get("year", ""),
            brand_logo=v.get("brand_logo", ""),
            image=v["image"],
        )
        for v in items
    ]


@router.get("/vehicles/{vehicle_id}", response_model=CatalogVehicle)
def get_vehicle_detail(vehicle_id: str):
    vehicle = get_vehicle(vehicle_id)
    if vehicle is None:
        raise HTTPException(status_code=404, detail="خودرو یافت نشد")
    return CatalogVehicle(**vehicle, categories=CATALOG_CATEGORIES)


@router.get("/vehicles/{vehicle_id}/views/{view_id}/hotspots", response_model=list[CatalogHotspot])
def get_view_hotspots(vehicle_id: str, view_id: str, category: Optional[str] = None):
    vehicle = get_vehicle(vehicle_id)
    if vehicle is None:
        raise HTTPException(status_code=404, detail="خودرو یافت نشد")
    if not any(v["id"] == view_id for v in vehicle["views"]):
        raise HTTPException(status_code=404, detail="نما یافت نشد")
    hotspots = get_hotspots(vehicle_id, view_id)
    if category:
        hotspots = [h for h in hotspots if h["category"] == category]
    return [CatalogHotspot(**h) for h in hotspots]


@router.get("/vehicles/{vehicle_id}/search-hotspots")
def search_vehicle_hotspots(vehicle_id: str, q: str = Query(..., min_length=1)):
    vehicle = get_vehicle(vehicle_id)
    if vehicle is None:
        raise HTTPException(status_code=404, detail="خودرو یافت نشد")
    return search_hotspots(vehicle_id, q)


@router.get("/hotspots/{vehicle_id}/{hotspot_id}/product", response_model=CatalogHotspotProduct)
def get_hotspot_product(vehicle_id: str, hotspot_id: str, view_id: str = Query(...)):
    hotspots = get_hotspots(vehicle_id, view_id)
    hotspot = next((h for h in hotspots if h["id"] == hotspot_id), None)
    if hotspot is None:
        raise HTTPException(status_code=404, detail="قطعه یافت نشد")

    conn = get_conn()
    row = conn.execute("SELECT * FROM products WHERE id = ?", (hotspot["product_id"],)).fetchone()
    if row is None:
        conn.close()
        raise HTTPException(status_code=404, detail="محصول یافت نشد")

    product = row_to_product(row)
    related = conn.execute(
        "SELECT * FROM products WHERE category = ? AND id != ? ORDER BY popularity DESC LIMIT 6",
        (product["category"], product["id"]),
    ).fetchall()
    conn.close()

    return CatalogHotspotProduct(
        hotspot=CatalogHotspot(**hotspot),
        product=Product(**product),
        part_number=hotspot.get("part_number", ""),
        oem_number=hotspot.get("oem", ""),
        material="پلاستیک / شیشه",
        weight_grams=1250,
        warranty="۶ ماه",
        related=[Product(**row_to_product(r)) for r in related],
    )

File: frontend/lib/models/smart_catalog.dart
import '../models/product.dart';

class CatalogVehicleSummary {
  final String id;
  final String name;
  final String subtitle;
  final String year;
  final String brandLogo;
  final String image;

  CatalogVehicleSummary({
    required this.id,
    required this.name,
    this.subtitle = '',
    this.year = '',
    this.brandLogo = '',
    required this.image,
  });

  factory CatalogVehicleSummary.fromJson(Map<String, dynamic> json) => CatalogVehicleSummary(
        id: json['id'] as String,
        name: json['name'] as String,
        subtitle: json['subtitle'] as String? ?? '',
        year: json['year'] as String? ?? '',
        brandLogo: json['brand_logo'] as String? ?? '',
        image: json['image'] as String,
      );
}

class CatalogView {
  final String id;
  final String name;
  final String image;

  CatalogView({required this.id, required this.name, required this.image});

  factory CatalogView.fromJson(Map<String, dynamic> json) => CatalogView(
        id: json['id'] as String,
        name: json['name'] as String,
        image: json['image'] as String,
      );
}

class CatalogCategory {
  final String id;
  final String name;
  final String icon;

  CatalogCategory({required this.id, required this.name, required this.icon});

  factory CatalogCategory.fromJson(Map<String, dynamic> json) => CatalogCategory(
        id: json['id'] as String,
        name: json['name'] as String,
        icon: json['icon'] as String,
      );
}

class CatalogHotspot {
  final String id;
  final String label;
  final String category;
  final double x;
  final double y;
  final int productId;
  final String partNumber;
  final String oem;

  CatalogHotspot({
    required this.id,
    required this.label,
    required this.category,
    required this.x,
    required this.y,
    required this.productId,
    this.partNumber = '',
    this.oem = '',
  });

  factory CatalogHotspot.fromJson(Map<String, dynamic> json) => CatalogHotspot(
        id: json['id'] as String,
        label: json['label'] as String,
        category: json['category'] as String,
        x: (json['x'] as num).toDouble(),
        y: (json['y'] as num).toDouble(),
        productId: json['product_id'] as int,
        partNumber: json['part_number'] as String? ?? '',
        oem: json['oem'] as String? ?? '',
      );
}

class CatalogVehicleDetail {
  final String id;
  final String name;
  final String subtitle;
  final String year;
  final String brandLogo;
  final String image;
  final List<CatalogView> views;
  final List<CatalogCategory> categories;

  CatalogVehicleDetail({
    required this.id,
    required this.name,
    this.subtitle = '',
    this.year = '',
    this.brandLogo = '',
    required this.image,
    required this.views,
    this.categories = const [],
  });

  factory CatalogVehicleDetail.fromJson(Map<String, dynamic> json) => CatalogVehicleDetail(
        id: json['id'] as String,
        name: json['name'] as String,
        subtitle: json['subtitle'] as String? ?? '',
        year: json['year'] as String? ?? '',
        brandLogo: json['brand_logo'] as String? ?? '',
        image: json['image'] as String,
        views: (json['views'] as List).map((e) => CatalogView.fromJson(e as Map<String, dynamic>)).toList(),
        categories: (json['categories'] as List? ?? [])
            .map((e) => CatalogCategory.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class CatalogHotspotProduct {
  final CatalogHotspot hotspot;
  final Product product;
  final String partNumber;
  final String oemNumber;
  final String material;
  final int weightGrams;
  final String warranty;
  final List<Product> related;

  CatalogHotspotProduct({
    required this.hotspot,
    required this.product,
    this.partNumber = '',
    this.oemNumber = '',
    this.material = '',
    this.weightGrams = 0,
    this.warranty = '',
    this.related = const [],
  });

  factory CatalogHotspotProduct.fromJson(Map<String, dynamic> json) => CatalogHotspotProduct(
        hotspot: CatalogHotspot.fromJson(json['hotspot'] as Map<String, dynamic>),
        product: Product.fromJson(json['product'] as Map<String, dynamic>),
        partNumber: json['part_number'] as String? ?? '',
        oemNumber: json['oem_number'] as String? ?? '',
        material: json['material'] as String? ?? '',
        weightGrams: json['weight_grams'] as int? ?? 0,
        warranty: json['warranty'] as String? ?? '',
        related: (json['related'] as List? ?? [])
            .map((e) => Product.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class HotspotSearchResult {
  final CatalogHotspot hotspot;
  final String viewId;

  HotspotSearchResult({required this.hotspot, required this.viewId});

  factory HotspotSearchResult.fromJson(Map<String, dynamic> json) => HotspotSearchResult(
        hotspot: CatalogHotspot.fromJson(json),
        viewId: json['view_id'] as String,
      );
}
